### ISSUE TITLE
Gensim Doc2Vec most_similar crashes when executed inside Celery Task

### ISSUE THREAD

**Issue by kalinkirev** [RaRe-Technologies/gensim#1293]
#### Description
I have and trained a doc2vec model. and a simple method that calls the most_similar function. When there is new text, i am inferring the vector from the trained model and then passing it on to most_similar function. It all works as expected. However, if i pass this function to Celery+Redis server it throws an exception. Please note that that Celery is Celery Flask_SQLAlchemy optimized object. Also, i'm running everything on my Mac, with Redis inside a Docker container. See below the code.

#### Steps/Code/Corpus to Reproduce
~Application/app/config/extensions.py
```
from celery import Celery

...

def make_celery(app):
    celery = Celery(app.import_name, backend=app.config['CELERY_RESULT_BACKEND'], broker=app.config['CELERY_BROKER_URL'])
    celery.conf.update(app.config)
    TaskBase = celery.Task

    class ContextTask(TaskBase):
        abstract = True

        def __call__(self, *args, **kwargs):
            with app.app_context():
                return TaskBase.__call__(self, *args, **kwargs)
    celery.Task = ContextTask
    return celery
```
~Application/app/__init__.py
```
from flask import Flask
from .config.extensions import make_celery

...

app = Flask(__name__)
celery = make_celery(app)
```

~/Application/app/ai/Similarity.py
```

from gensim.models import Doc2Vec
from app import celery



doc2vec_clf = Doc2Vec.load('text_similarity.doc2vec')

@celery.task(bind=True)
def most_similar(self, text, topN=10):
    try:
        doc2vec_clf = TicketSimilarityClassifier()
        test_vector = doc2vec_clf.infer_vector(utils.simple_preprocess(fix_text(text)))
        app.logger.debug("Test Vector from text:<%s> has been infered as: <%s>" % (text, test_vector))
        return dict(doc2vec_clf.docvecs.most_similar([test_vector], topn=topN))
    except Exception as e:
        app.logger.warning("Could not find similar entries to ticket %s. Could not infer vector" % (text), exc_info=True)
        return None

test_text = u'I lost my email keys please help'
```

~/Application/app/ai/test.py
```
from .Similarity import most_similar

#Working example
res1 = most_similar(test_text)
print res1

#Not working 
res2 = most_similar.delay(test_text)
print res2
```

#### Expected Results

This is output from the logger
```
2017-04-26 18:13:02,045:DEBUG:/Users/kalinkirev/Documents/Python/Projects/Application/app/ai/Similarity.py:84:most_similar():Test Vector from text:<I lost my email keys please help> has been inferred as: 
<[-0.14122523  0.24877205  0.12713383 -0.36630598 -0.04363676  0.13560387
 -0.12484436  0.02849481 -0.07546753  0.07372763 -0.10689059 -0.00361574
 -0.07099307  0.10793268  0.19399352 -0.04166874  0.02890538  0.13150787
  0.07709999 -0.05014185 -0.18250267 -0.14677131  0.0974682   0.03343132
  0.01735616  0.10136632 -0.04067229  0.22774744  0.18592837  0.18922505
 -0.08336212  0.10215112  0.19345225 -0.0412197   0.05416567  0.09264339
 -0.03123957  0.00410211 -0.20401753  0.05664713 -0.03602515 -0.10209432
  0.05511197  0.01955108 -0.0791814   0.04744654 -0.07834043 -0.09010441
 -0.04490344 -0.04122124 -0.02422909 -0.16243012 -0.04704344  0.24701518
 -0.04385039 -0.06432433 -0.00738273 -0.16356464  0.14755113  0.19551882
  0.1172388   0.11033925  0.21704715  0.33036557  0.02260509  0.05479171
 -0.13214304 -0.07105681 -0.14401715  0.02182562 -0.24489661  0.07379169
  0.18064617 -0.0494834  -0.04035234 -0.17701541  0.1004746   0.00889999
  0.12171029  0.2621589   0.20764789 -0.03447072 -0.20462507  0.11168189
 -0.06970863  0.02121511 -0.0398889  -0.0004392  -0.05589946  0.06230953
  0.00753125  0.02450103 -0.06608684  0.02616308  0.14001457  0.09255141
 -0.05373417 -0.06442818 -0.06763304 -0.03849017]>
```
This is the output from print with: most_similar(test_text)
```
{'T20160505.1165': 0.8388196229934692,
 'T20160506.0494': 0.8388283252716064,
 'T20160506.0845': 0.8485631346702576,
 'T20160510.1240': 0.8523017168045044,
 'T20160511.0816': 0.8528236150741577,
 'T20160511.0940': 0.8413317799568176,
 'T20160512.0938': 0.8459789156913757,
 'T20160513.0848': 0.8478519916534424,
 'T20160517.0108': 0.8394412994384766,
 'T20160519.1516': 0.8446635007858276}
```

#### Actual Results

This is the output from Celery console with: most_similar.delay(test_text)
```
[2017-04-26 18:14:41,773: INFO/MainProcess] Connected to redis://localhost:6379//
[2017-04-26 18:14:41,789: INFO/MainProcess] mingle: searching for neighbors
[2017-04-26 18:14:42,830: INFO/MainProcess] mingle: all alone
[2017-04-26 18:14:42,901: INFO/MainProcess] celery@smapp ready.
[2017-04-26 18:15:09,337: INFO/MainProcess] Received task: app.ai.TicketSimilarity.most_similar[4549bd27-7757-476c-bf16-69cbfb7deb65]  
[2017-04-26 18:15:09,345: INFO/PoolWorker-2] loading Doc2Vec object from /Users/kalinkirev/Documents/Python/Projects/smapp/Application/app/ai/text_similarity.doc2vec
[2017-04-26 18:15:09,666: INFO/PoolWorker-2] loading docvecs recursively from /Users/kalinkirev/Documents/Python/Projects/smapp/Application/app/ai/text_similarity.doc2vec.* with mmap=None
[2017-04-26 18:15:09,667: INFO/PoolWorker-2] setting ignored attribute syn0norm to None
[2017-04-26 18:15:09,667: INFO/PoolWorker-2] setting ignored attribute cum_table to None
[2017-04-26 18:15:09,668: INFO/PoolWorker-2] loaded /Users/kalinkirev/Documents/Python/Projects/smapp/Application/app/ai/text_similarity.doc2vec
[2017-04-26 18:15:09,756: INFO/PoolWorker-2] precomputing L2-norms of doc weight vectors

[2017-04-26 18:15:09,785: ERROR/MainProcess] Process 'PoolWorker-2' pid:5536 exited with 'signal 11 (SIGSEGV)'
[2017-04-26 18:15:09,876: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 11 (SIGSEGV).',)
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/billiard/pool.py", line 1224, in mark_as_worker_lost
    human_status(exitcode)),
WorkerLostError: Worker exited prematurely: signal 11 (SIGSEGV).
```

This is the output from the Logger
Logger output shows that the vector has been inferred successfully. See timestamps
```
2017-04-26 18:15:09,725:DEBUG:/Users/kalinkirev/Documents/Python/Projects/Application/app/ai/Similarity.py:84:most_similar():Test Vector from text:<I lost my email keys please help> has been inferred as: 
<[-0.14122523  0.24877205  0.12713383 -0.36630598 -0.04363676  0.13560387
 -0.12484436  0.02849481 -0.07546753  0.07372763 -0.10689059 -0.00361574
 -0.07099307  0.10793268  0.19399352 -0.04166874  0.02890538  0.13150787
  0.07709999 -0.05014185 -0.18250267 -0.14677131  0.0974682   0.03343132
  0.01735616  0.10136632 -0.04067229  0.22774744  0.18592837  0.18922505
 -0.08336212  0.10215112  0.19345225 -0.0412197   0.05416567  0.09264339
 -0.03123957  0.00410211 -0.20401753  0.05664713 -0.03602515 -0.10209432
  0.05511197  0.01955108 -0.0791814   0.04744654 -0.07834043 -0.09010441
 -0.04490344 -0.04122124 -0.02422909 -0.16243012 -0.04704344  0.24701518
 -0.04385039 -0.06432433 -0.00738273 -0.16356464  0.14755113  0.19551882
  0.1172388   0.11033925  0.21704715  0.33036557  0.02260509  0.05479171
 -0.13214304 -0.07105681 -0.14401715  0.02182562 -0.24489661  0.07379169
  0.18064617 -0.0494834  -0.04035234 -0.17701541  0.1004746   0.00889999
  0.12171029  0.2621589   0.20764789 -0.03447072 -0.20462507  0.11168189
 -0.06970863  0.02121511 -0.0398889  -0.0004392  -0.05589946  0.06230953
  0.00753125  0.02450103 -0.06608684  0.02616308  0.14001457  0.09255141
 -0.05373417 -0.06442818 -0.06763304 -0.03849017]>
```

#### Versions
NOTE: I have created the trained model in an earlier version 0.13.2. But experienced the same issue. Then I updated gensim but still no success. 
```
Python 2.7.9 (default, May 25 2015, 11:14:22) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform; print(platform.platform())
Darwin-16.5.0-x86_64-i386-64bit
>>> import sys; print("Python", sys.version)
('Python', '2.7.9 (default, May 25 2015, 11:14:22) \n[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)]')
>>> import numpy; print("NumPy", numpy.__version__)
('NumPy', '1.12.1')
>>> import scipy; print("SciPy", scipy.__version__)
('SciPy', '0.19.0')
>>> import gensim; print("gensim", gensim.__version__)
('gensim', '1.0.1')
>>> from gensim.models import word2vec;print("FAST_VERSION", word2vec.FAST_VERSION)
('FAST_VERSION', 0)
```

Thank you

**Comment by gojomo** [comment]
Are you sure the error is occurring in `most_similar()`? You may want to split the `return` line to do the `most_similar()` separately, and print debug info if it's completed, to be sure that's where the segfault occurs. 

How large is your model, and what does memory usage look like on the worker-machine, just before the operation is attempted? Can you test with a tiny model, and see if the same error occurs?

Note that `most_similar()` requires the `doctag_syn0norm` array of unit-normalized vectors to be available, so the 1st call will calculate those – roughly doubling memory use at that moment. (Then, the calculation of all pairwise distances for the specific call will temporarily use a bunch more memory.) Maybe, what's really an out-of-addressable-memory problem is showing up as a segfault. 

You could force the unit-normed vectors to be precalculated before the first call with `doc2vec_clf.docvecs.init_sims()`. If you're only doing these most-similiar ops, and thus only need the unit-normed vectors, you can save memory by clobbering the original raw-magnitude vectors in-place by instead using `doc2vec_clf.docvecs.init_sims(replace=True)`.

**Comment by kalinkirev** [comment]
Thank you for your quick reply. My model is relatively small about 20mb.  First I did try your suggestion and ran ```doc2ve_clf.docvecs.init_sims(replace=True)```, but that still didn't work out. 

Then, I ran [Doc2Vec Tutorial on the Lee Dataset.ipynb](https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/doc2vec-lee.ipynb) as a separate project. Trained the model and wrapped it in a Celery.Task, and I had no issues. 

Afterwards, I took that same Lee Model and try to wrap it within my Flask task but was still getting an error. 

As a last attempt I created a very cutdown version of my Flask app (pretty much including only the Gensim stuff) and it was working. 

So, at this stage i'm assuming my application footprint is a bit big. So I will be looking for a proper way to decouple those backend components from the main Application servers. Thus, I consider the issue as closed as it does not appear to be gensim related. Thanks for the hard work! 

P.S. Its very strange the in the original design all other Celery tasks where being executed without issues, but the only one that was failing was the one with Gensim...
